test/e2e/app_suffix_test.go

```diff
@@ -0,0 +1,146 @@
+metadata:
+  name: test-app-suffix-app-exists
+  labels:
+    kapp.k14s.io/is-app: ""
+data:
+  spec: '{"labelKey":"kapp.k14s.io/app","labelValue":"1641592268838201000","lastChange":{"startedAt":"0001-01-01T00:00:00Z","finishedAt":"0001-01-01T00:00:00Z"}}'
+`
+
+	yaml1 := `
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: redis-primary
+spec:
+  ports:
+  - port: 6380
+    targetPort: 6380
+  selector:
+    app: redis
+    tier: backend
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: redis-config
+data:
+  key: value
+`
+
+	yaml2 := `
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: redis-config
+data:
+  key: value2
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: redis-config2
+data:
+  key: value
+`
+
+	logger.Section("deploy with 1 delete, 1 update, 1 create", func() {
+		kubectl.RunWithOpts([]string{"apply", "-f", "-"},
+			RunOpts{IntoNs: true, StdinReader: strings.NewReader(existingApp)})
+		NewPresentClusterResource("configmap", name, env.Namespace, kubectl)
+
+		kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name}, RunOpts{IntoNs: true, StdinReader: strings.NewReader(yaml1)})
+		NewPresentClusterResource("configmap", name+app.AppSuffix, env.Namespace, kubectl)
+		NewMissingClusterResource(t, "configmap", name, env.Namespace, kubectl)
+
+		kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name}, RunOpts{IntoNs: true, StdinReader: strings.NewReader(yaml2)})
+		NewPresentClusterResource("configmap", name+app.AppSuffix, env.Namespace, kubectl)
+		NewMissingClusterResource(t, "configmap", name, env.Namespace, kubectl)
+
+		cleanUp()
+	})
+
+	logger.Section("rename", func() {
+		newName := "test-app-suffix-app-exists-new"
+
+		kubectl.RunWithOpts([]string{"apply", "-f", "-"},
+			RunOpts{IntoNs: true, StdinReader: strings.NewReader(existingApp)})
+		NewPresentClusterResource("configmap", name, env.Namespace, kubectl)
+
+		kapp.Run([]string{"rename", "-a", name, "--new-name", newName})
+		NewPresentClusterResource("configmap", newName+app.AppSuffix, env.Namespace, kubectl)
+
+		kapp.Run([]string{"delete", "-a", newName})
+	})
+
+	logger.Section("delete", func() {
+		kubectl.RunWithOpts([]string{"apply", "-f", "-"},
+			RunOpts{IntoNs: true, StdinReader: strings.NewReader(existingApp)})
+		NewPresentClusterResource("configmap", name, env.Namespace, kubectl)
+
+		kapp.Run([]string{"delete", "-a", name})
+
+		NewMissingClusterResource(t, "configmap", name, env.Namespace, kubectl)
+	})
+
+	logger.Section("configmap with suffix exists and not marked as a kapp-app", func() {
+		fqName := name + app.AppSuffix
+
+		NewClusterResource(t, "configmap", fqName, env.Namespace, kubectl)
+		NewPresentClusterResource("configmap", fqName, env.Namespace, kubectl)
+
+		_, err := kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name},
+			RunOpts{IntoNs: true, AllowError: true, StdinReader: strings.NewReader(yaml1)})
+
+		require.Containsf(t, err.Error(), "kapp: Error:", "did not contain parseable app metadata")
```

Comment:
Something to note from the issue https://github.com/vmware-tanzu/carvel-kapp/issues/146 was my suggestion that:

```
Given: Existing application named: foo and configmap: foo.apps.k14s.io

if foo.apps.k14s.io already exists and not marked as a kapp app configmap, error and prevent the deploy continuing. Suggest to the user that the configmap should be renamed / use the environment variable USE_EXISTING_CONFIGMAP_NAME=true to prevent migratation to the newly generated configmap name.
```

The existing behaviour of kapp was to fail with a "cannot parse" error if someone had manually created a configmap that matched the app name. I found it hard to find a good place for this error such that it only happened during deploy, as kapp in other methods calls [setMeta()](https://github.com/vmware-tanzu/carvel-kapp/blob/develop/pkg/kapp/app/recorded_app.go#L227-L238) before [CreateOrUpdate()](https://github.com/vmware-tanzu/carvel-kapp/blob/develop/pkg/kapp/app/recorded_app.go#L80) even runs

Reply:
One consequence of having it there is the following:

Example Workflow:
1. `kubectl create configmap neil.apps.k14s.io` // pre-existing non-kapp configmap
2. `kapp deploy -f examples/jobs/successful-job.yml -a neil`

> kapp: Error: ConfigMap 'neil.apps.k14s.io' already exists and did not contain parseable app metadata: (hint: rename / delete this configmap or set the environment variable USE_EXISTING_CONFIGMAP_NAME=true to prevent migration)

BUT also the other command render this error, which doesn't feel right as the hint isn't really true. The env var will do nothing to help in these cases:

1. `kapp inspect -a neil`  
2. `kapp delete  -a neil`

> kapp: Error: ConfigMap 'neil.apps.k14s.io' already exists and did not contain parseable app metadata: (hint: rename / delete this configmap or set the environment variable USE_EXISTING_CONFIGMAP_NAME=true to prevent migration)

---

I am not sure what the "right" thing is here, the problem lies in the fact all commands check that the app exists / read the metadata, so whatever error we throw in `meta()` or `exists()` will show up for all commands `deploy` , `inspect` and `delete`. Ideally this error only shows up during a `deploy` and not the other commands, but I don't think that can happen, as these other functions are always called before `CreateOrUpdate()`

Reply:
I am actually comfortable with the existing error as well, as it would we be very rare and we would also have it documented.